### ISSUE TITLE
[no-unused-vars] Getting false positives in typescript files when using `?` optional chaining syntax

### ISSUE THREAD

**Issue by makesitgo** [typescript-eslint/typescript-eslint#3385]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

Hard to tell how to repro, because we have this rule both working and not working through the codebase.  The title here is mostly a theory, but of the 53 instances of incorrect `no-unused-vars` I am so far consistently seeing usages of `?` present in that same line of code.

For instance:

```ts
function mapStateToProps(state: RootState) {
  // linter complains: 'appMetricsById' is assigned a value but never used
  const { appMetricsById } = getMeasurementsState(state);
  return {
    measurements: appMetricsById[appId]?.measurementsByName || DEFAULT_MEASUREMENTS,
  };
}
```

Another example (with a component prop):

```ts
// linter complains: 'opt' is defined but never used. Allowed unused args must match /^_/u
onChange={(opt?: CreatablePartitionOption) => onPartitionChange(opt?.value)}
```

Last one:

```ts
const mapStateToProps = (state: RootState, ownProps: OwnProps) => {
  // linter complains: 'match' is assigned a value but never used
  const { match } = ownProps;

  return {
    editingSecret: {
      isLoading: secrets.isLoading || editingSecret.isLoading,
      data: editingSecret.data || secrets.data?.find((secret) => secret.id === match.params.id) || null,
      error: secrets.error || editingSecret.error,
    }
  };
};
```

All of these vars that are actually used seem to be on lines where we use typescript's [optional chaining](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html) syntax

```js
{
  "extends": [
    ...
    'plugin:@typescript-eslint/recommended',
    ...
  ],
  "parser": 'typescript-eslint/parser',
  "plugins": [
    ...
    '@typescript-eslint',
    ...
  ],
  "rules": {
    ...
    'no-unused-vars': 0,
    '@typescript-eslint/no-unused-vars': [2, { argsIgnorePattern: '^_' }],
    ....
  }
}
```

**Expected Result**

I would expect these vars to be checked as used.

**Actual Result**

These vars are erroneously reported as being unused.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `^4.23.0` |
| `@typescript-eslint/parser`        | `^4.23.0` |
| `TypeScript`                       | `^4.2.4` |
| `ESLint`                           | `^7.26.0` |
| `node`                             | `12.16.2` |

> NOTE: `node` is not a package dependency, that version is just what's installed on my local machine

**Comment by bradzacher** [comment]
unable to repro.
![image](https://user-images.githubusercontent.com/7462525/118175828-7662c800-b3e5-11eb-8fa2-7c220a538ef0.png)

Highly likely that you have the wrong version(s) of packages installed.

**Comment by makesitgo** [comment]
Odd... I actually stumbled across this while doing a project-wide dependency update, and these eslint/typescript dependencies were my last ones (so more or less things should "current" as of a few days ago).  do any of the versions I listed above jump out to you as incorrect?

I'll try blowing away `node_modules` see where the dust settles, glad to see it working on your end though

**Comment by makesitgo** [comment]
🎉  ... that did the trick and the errors have gone away.  thanks for the sanity check, apologies for the noise!
